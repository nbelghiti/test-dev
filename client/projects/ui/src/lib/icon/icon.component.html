<i class="c-icon__glyph {{ size && 'c-icon__glyph--' + size }} {{ type && 'c-icon__glyph--' + type }} fas fa-{{ name }}"></i>
