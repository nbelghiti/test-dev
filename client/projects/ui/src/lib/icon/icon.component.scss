// =============================================================================
// :: Imports
// =============================================================================
@import "global";

// =============================================================================
// :: Icon
// =============================================================================
:host {
	display: inline-flex;
	justify-content: center;
	align-items: center;

	font-size: inherit;
}

// =============================================================================
// :: Elements
// =============================================================================
.c-icon__glyph {
	line-height: inherit;

	cursor: pointer;
}

// =============================================================================
// :: Modifiers
// =============================================================================
.c-icon__glyph--large {
	font-size: 2rem;
}

.c-icon__glyph--small {
	font-size: 1.4rem;
}

.c-icon__glyph--tiny {
	font-size: 1.2rem;
}

.c-icon__glyph--success {
	color: $ui-color-success-dark;
}

.c-icon__glyph--error {
	color: $ui-color-error-dark;
}

.c-icon__glyph--warning {
	color: $ui-color-warning-dark;
}

.c-icon__glyph--info {
	color: $ui-color-info-dark;
}
