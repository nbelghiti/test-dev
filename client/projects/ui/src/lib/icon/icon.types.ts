export enum IconName {
	archive = 'archive',
	arrowDown = 'arrow-down',
	arrowLeft = 'arrow-left',
	arrowRight = 'arrow-right',
	arrowRightAlt = 'arrow-alt-circle-right',
	arrows = 'arrows-alt-h',
	arrowUp = 'arrow-up',
	bell = 'bell',
	bellSlash = 'bell-slash',
	book = 'book',
	brackets = 'brackets',
	calendar = 'calendar',
	caretDown = 'caret-down',
	caretLeft = 'caret-left',
	caretRight = 'caret-right',
	caretUp = 'caret-up',
	check = 'check',
	checkCircle = 'check-circle',
	chevronDown = 'chevron-down',
	chevronLeft = 'chevron-left',
	chevronRight = 'chevron-right',
	chevronUp = 'chevron-up',
	clock = 'clock',
	clone = 'clone',
	close = 'times-circle',
	code = 'code',
	codeBranch = 'code-branch',
	cube = 'cube',
	download = 'download',
	edit = 'edit',
	excel = 'file-excel',
	exclamation = 'exclamation',
	exclamationCircle = 'exclamation-circle',
	expand = 'expand',
	externalLink = 'external-link-alt',
	file = 'file-alt',
	home = 'home',
	horizontalEllipsis = 'ellipsis-h',
	industry = 'industry',
	infoCircle = 'info-circle',
	invisible = 'eye-slash',
	marker = 'map-marker-alt',
	minus = 'minus',
	pdf = 'file-pdf',
	pen = 'pen',
	plus = 'plus',
	print = 'print',
	search = 'search',
	settings = 'cog',
	signOut = 'sign-out-alt',
	snowflake = 'snowflake',
	spinner = 'spinner',
	star = 'star',
	sync = 'sync-alt',
	table = 'table',
	times = 'times',
	trash = 'trash',
	upload = 'upload',
	user = 'user',
	users = 'users',
	visible = 'eye',
	plusSquare = 'plus-square',
	minusSquare = 'minus-square',
}

export enum IconType {
	success = 'success',
	error = 'error',
	warning = 'warning',
	info = 'info',
}

export enum IconSize {
	large = 'large',
	small = 'small',
	tiny = 'tiny',
}
