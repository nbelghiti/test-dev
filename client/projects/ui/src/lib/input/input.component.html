<div
	class="c-input-field"
	[ngClass]="{
		'c-input-field--disabled': value.disabled,
		'c-input-field--error': error
	}"
>
	<ui-icon
		*ngIf="icon"
		[name]="icon"
		class="c-input-field__icon"
	>
	</ui-icon>

	<div class="c-input-field__content">
		<div class="c-input-field__input">
			<input
				[id]="name"
				[name]="name"
				[type]="type"
				class="c-input-field__field"
				[placeholder]="placeholder"
				[formControl]="value"
				required
			/>

			<label
				class="c-input-field__label"
				[for]="name"
			>
				{{label}}
			</label>
		</div>

		<div *ngIf="error" class="c-error">
			<ui-icon
				name="exclamation-circle"
				size="small"
			>
			</ui-icon>

			<span class="c-error__text">
				{{error}}
			</span>
		</div>
	</div>
</div>
