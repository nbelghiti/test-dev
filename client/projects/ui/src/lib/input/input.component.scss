// =============================================================================
// :: Imports
// =============================================================================
@import "global";

// =============================================================================
// :: Input field
// =============================================================================
.c-input-field {
	display: flex;
	align-items: center;

	background-color: $background-color-base;
	border: $border-width-base solid $border-color-grey-76;

	& + .c-input-field {
		margin-top: -0.1rem;
	}
}

// =============================================================================
// :: Elements
// =============================================================================
.c-input-field__content,
.c-input-field__input {
	position: relative;

	width: 100%;
}

.c-input-field__icon {
	margin: 0 1rem;
}

.c-input-field__label {
	position: absolute;
	top: 50%;
	left: 1rem;

	order: -1;

	font-size: $font-size-1x-small;
	line-height: $line-height-base;

	color: $text-color-grey-56;

	transform-origin: left center;
	transform: scale(1.15) translateY(-50%);

	transition: transform $animation-base;

	pointer-events: none;
}

.c-input-field__field {
	display: inline-block;

	width: 100%;
	padding: 2.4rem 2rem 0.2rem 1rem;

	font-size: $font-size-base;
	line-height: $line-height-base;

	background: none;

	color: $text-color-base;

	border: $border-width-base solid transparent;

	transition:
		border $animation-base,
		box-shadow $animation-base;

	cursor: text;

	&:focus {
		outline: none;
	}

	&::placeholder {
		color: $text-color-grey-56;

		opacity: 0;

		transition: all $animation-base;

		visibility: hidden;
	}

	&:focus,
	&:valid {

		+ .c-input-field__label {
			transform: scale(1) translateY(-100%);
		}

		&::placeholder {

			opacity: 1;

			visibility: visible;
		}
	}

	&:disabled {

		+ .c-input-field__label {
			transform: scale(1) translateY(-100%);
		}
	}
}

// =============================================================================
// :: Modifiers
// =============================================================================
.c-input-field--disabled {
	cursor: not-allowed;

	.c-input-field__field {
		background-color: $background-color-grey-96;
		color: $text-color-grey-46;

		pointer-events: none;
	}

	.c-input-field__label {
		cursor: not-allowed;
	}
}

.c-input-field--error {

	.c-input-field__field {
		width: calc(100% - 3rem);

		margin: 2.4rem 2rem 0.2rem 1rem;
		padding: 0;

		border: $border-width-base solid $ui-color-error-dark;
	}

	.c-input-field__label {
		transform: scale(1) translateY(-100%);
	}

	.c-error {
		padding: 0 0 0.2rem 1rem;
	}
}
