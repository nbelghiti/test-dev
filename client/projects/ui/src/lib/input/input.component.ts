import { Component, Input, OnDestroy, forwardRef } from '@angular/core';
import { ControlValueAccessor, FormControl, NG_VALUE_ACCESSOR } from '@angular/forms';
import { Subject } from 'rxjs';
import { takeUntil } from 'rxjs/operators';

@Component({
	selector: 'ui-input',
	templateUrl: './input.component.html',
	styleUrls: ['./input.component.scss'],
	providers: [
		{
			provide: NG_VALUE_ACCESSOR,
			useExisting: forwardRef(() => InputComponent),
			multi: true,
		},
	],
})

export class InputComponent implements ControlValueAccessor, OnDestroy {
	@Input() public name: string;
	@Input() public placeholder: string = '';
	@Input() public label: string;
	@Input() public type?: string = 'text';
	@Input() public error?: string = '';
	@Input() public icon?: string = '';

	public value: FormControl = new FormControl(null);

	public onChange: (value: string) => void;
	public onTouched: () => void;

	private destroyed$: Subject<boolean> = new Subject<boolean>();

	constructor() {
		this.value.valueChanges
			.pipe(
				takeUntil(this.destroyed$),
			)
			.subscribe((value: string) => {
				this.onChange(value);
			});
	}

	public ngOnDestroy(): void {
		this.destroyed$.next(true);
		this.destroyed$.complete();
	}

	public writeValue(value: string): void {
		this.value.setValue(value, { emitEvent: false });
	}

	public registerOnChange(onChange: (_) => any): void {
		this.onChange = onChange;
	}

	public registerOnTouched(fn: () => void): void {
		this.onTouched = fn;
	}

	public setDisabledState(isDisabled: boolean): void {
		if (isDisabled) {
			this.value.disable();
		} else {
			this.value.enable();
		}
	}
}
