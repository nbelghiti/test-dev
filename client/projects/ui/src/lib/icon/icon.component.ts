import { Component, HostBinding, Input } from '@angular/core';

import { IconName, IconSize, IconType } from './icon.types';

@Component({
	selector: 'ui-icon',
	templateUrl: './icon.component.html',
	styleUrls: ['./icon.component.scss']
})

export class IconComponent {
	@Input() public name: IconName;
	@Input() public type: IconType = null;
	@Input() public size: IconSize = null;

	@HostBinding('class.c-icon') public className = true;
}
