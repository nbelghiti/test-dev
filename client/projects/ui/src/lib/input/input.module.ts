import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';

import { IconModule } from '../icon/icon.module';

import { InputComponent } from './input.component';

@NgModule({
	imports: [
		CommonModule,
		IconModule,
		FormsModule,
		ReactiveFormsModule,
	],
	declarations: [
		InputComponent,
	],
	exports: [
		InputComponent,
	],
})
export class InputModule {}
