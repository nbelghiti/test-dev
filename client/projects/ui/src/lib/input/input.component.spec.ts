import { ComponentFixture, TestBed, async } from '@angular/core/testing';
import { ReactiveFormsModule } from '@angular/forms';
import { By } from '@angular/platform-browser';

import { IconComponent } from '../icon/icon.component';

import { InputComponent } from './input.component';

describe('InputComponent', () => {
	let component: InputComponent;
	let fixture: ComponentFixture<InputComponent>;

	beforeEach(async(() => {
		TestBed.configureTestingModule({
			imports: [
				ReactiveFormsModule,
			],
			declarations: [
				InputComponent,
				IconComponent
			]
		})
			.compileComponents();
	}));

	beforeEach(() => {
		fixture = TestBed.createComponent(InputComponent);
		component = fixture.componentInstance;
		fixture.detectChanges();
	});

	it('should create', () => {
		expect(component).toBeTruthy();
	});

	it('Should render correctly', () => {
		expect(fixture).toMatchSnapshot();
	});

	it('should set value to input after writeValue has been called', () => {
		const value = 'value';
		component.writeValue(value);

		fixture.detectChanges();

		const pElement = fixture.debugElement.query(By.css('input'));

		expect(pElement.nativeElement.value).toContain(value);
	});

	it('should set the disabled state of the input', () => {
		component.onChange = jest.fn();
		component.setDisabledState(true);

		fixture.detectChanges();

		const pElement = fixture.debugElement.query(By.css('input'));

		expect(pElement.nativeElement.disabled).toBe(true);

		component.setDisabledState(false);

		expect(pElement.nativeElement.disabled).toBe(false);
	});

	describe('destroy', () => {
		beforeEach(() => {
			fixture = TestBed.createComponent(InputComponent);
			component = fixture.componentInstance;
			fixture.detectChanges();
		});

		it('should set a value to the destroyed$ Subject, destroying all subscriptions', async (done: jest.DoneCallback) => {
			const testSub = component['destroyed$']
				.subscribe((triggered: boolean) => {
					expect(triggered).toBe(true);

					testSub.unsubscribe();

					done();
				});

			component.ngOnDestroy();
		});
	});
});
