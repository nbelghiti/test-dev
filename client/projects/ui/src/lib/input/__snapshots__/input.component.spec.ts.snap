// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputComponent Should render correctly 1`] = `
<ui-input
  destroyed$={[Function Subject]}
  error=""
  icon=""
  placeholder=""
  type={[Function String]}
  value={[Function FormControl]}
>
  <div
    class="c-input-field"
    ng-reflect-klass="c-input-field"
    ng-reflect-ng-class="[object Object]"
  >
    
    <div
      class="c-input-field__content"
    >
      <div
        class="c-input-field__input"
      >
        <input
          class="c-input-field__field ng-untouched ng-pristine ng-invalid"
          id="undefined"
          name="undefined"
          ng-reflect-form="[object Object]"
          ng-reflect-required=""
          placeholder=""
          required=""
          type="text"
        />
        <label
          class="c-input-field__label"
          for="undefined"
        >
            
        </label>
      </div>
      
    </div>
  </div>
</ui-input>
`;
